svm/tests/concurrent_tests.rs

```diff
@@ -97,3 +117,178 @@ fn test_program_cache_with_exhaustive_scheduler() {
     let runner = Runner::new(scheduler, Default::default());
     runner.run(move || program_cache_execution(4));
 }
+
+// This test executes multiple transactions in parallel where all read from the same data account,
+// but write to different accounts. Given that there are no locks in this case, SVM must behave
+// correctly.
+fn svm_concurrent() {
+    let mock_bank = Arc::new(MockBankCallback::default());
+    let batch_processor = Arc::new(TransactionBatchProcessor::<MockForkGraph>::new(
+        5,
+        2,
+        HashSet::new(),
+    ));
+    let fork_graph = Arc::new(RwLock::new(MockForkGraph {}));
+
+    create_executable_environment(
+        fork_graph.clone(),
+        &mock_bank,
+        &mut batch_processor.program_cache.write().unwrap(),
+    );
+
+    batch_processor.fill_missing_sysvar_cache_entries(&*mock_bank);
+    register_builtins(&mock_bank, &batch_processor);
+
+    let mut transaction_builder = SanitizedTransactionBuilder::default();
+    let program_id = deploy_program("transfer-from-account".to_string(), 0, &mock_bank);
+
+    const THREADS: usize = 4;
+    const TRANSACTIONS_PER_THREAD: usize = 3;
+    const AMOUNT: u64 = 50;
+    const CAPACITY: usize = THREADS * TRANSACTIONS_PER_THREAD;
+    const BALANCE: u64 = 500000;
+
+    let mut transactions = vec![Vec::new(); THREADS];
+    let mut check_data = vec![Vec::new(); THREADS];
+    let read_account = Pubkey::new_unique();
+    let mut account_data = AccountSharedData::default();
+    account_data.set_data(AMOUNT.to_le_bytes().to_vec());
+    mock_bank
+        .account_shared_data
+        .write()
+        .unwrap()
+        .insert(read_account, account_data);
+
+    #[derive(Clone)]
+    struct CheckTxData {
+        sender: Pubkey,
+        recipient: Pubkey,
+        fee_payer: Pubkey,
+    }
+
+    for idx in 0..CAPACITY {
+        let sender = Pubkey::new_unique();
+        let recipient = Pubkey::new_unique();
+        let fee_payer = Pubkey::new_unique();
+        let system_account = Pubkey::from([0u8; 32]);
+
+        let mut account_data = AccountSharedData::default();
+        account_data.set_lamports(BALANCE);
+
+        mock_bank
+            .account_shared_data
+            .write()
+            .unwrap()
+            .insert(sender, account_data.clone());
+        mock_bank
+            .account_shared_data
+            .write()
+            .unwrap()
+            .insert(recipient, account_data.clone());
+        mock_bank
+            .account_shared_data
+            .write()
+            .unwrap()
+            .insert(fee_payer, account_data);
```

Comment:
nit: I think you can just unlock this once and do all the writes before dropping the unlock.